File: ToDoModel/Dimension Libraries/ID/ListOfIDs.vb
Imports System.IO
Imports System.Numerics
Imports System.Runtime.Serialization.Formatters.Binary
Imports Microsoft.Office.Interop.Outlook
Imports UtilitiesVB

<Serializable()>
Public Class ListOfIDs
    Implements IListOfIDs

    Public UsedIDList As List(Of String)
    Private PMaxIDLength As Long
    Public pFileName As String = ""

    Public Sub New(ByVal listUsedID As List(Of String))
        UsedIDList = listUsedID
    End Sub

    Public Sub RePopulate(Application As Application) Implements IListOfIDs.RePopulate
        Dim unused As New Object
        Dim DM As New TreeOfToDoItems
        Dim ToDoList As List(Of Object) = DM.GetToDoList(TreeOfToDoItems.LoadOptions.vbLoadAll, Application)
        UsedIDList = New List(Of String)

        Dim ObjItem As Object
        For Each ObjItem In ToDoList
            Dim strID As String = CustomFieldID_GetValue(ObjItem, "ToDoID")
            If UsedIDList.Contains(strID) = False And strID.Length <> 0 Then
                UsedIDList.Add(strID)
                If strID.Length > PMaxIDLength Then PMaxIDLength = strID.Length
            End If
        Next
    End Sub

    'Public Sub CondenseIDs()
    '    Dim DM As DataModel_ToDoTree = New DataModel_ToDoTree
    '    DM.LoadTree(DataModel_ToDoTree.LoadOptions.vbLoadAll)

    '    Dim ToDoTree As List(Of TreeNode(Of ToDoItem)) = DM.ListOfToDoTree
    'End Sub

    Public ReadOnly Property MaxIDLength As Long Implements IListOfIDs.MaxIDLength
        Get
            If PMaxIDLength = 0 Then
                Dim maxLen As Long = 0
                For Each strID As String In UsedIDList
                    If strID.Length > maxLen Then
                        maxLen = strID.Length
                    End If
                Next
                PMaxIDLength = maxLen
            End If
            Return PMaxIDLength

        End Get
    End Property

    Public Function GetNextAvailableToDoID(strSeed As String) As String Implements IListOfIDs.GetNextAvailableToDoID
        Dim blContinue As Boolean = True
        Dim lngMaxID As BigInteger = ConvertToDecimal(125, strSeed)
        Dim strMaxID As String = ""

        While blContinue
            lngMaxID += 1
            strMaxID = ConvertToBase(125, lngMaxID)
            If UsedIDList.Contains(strMaxID) = False Then
                blContinue = False
            End If
        End While
        UsedIDList.Add(strMaxID)
        If strMaxID.Length > PMaxIDLength Then
            PMaxIDLength = strMaxID.Length
            My.Settings.MaxIDLength = PMaxIDLength
            My.Settings.Save()
        End If
        Return strMaxID
    End Function

    Public Function GetMaxToDoID() As String Implements IListOfIDs.GetMaxToDoID
        Dim strMaxID = UsedIDList.Max()
        Dim lngMaxID As BigInteger = ConvertToDecimal(125, strMaxID)
        lngMaxID += 1
        strMaxID = ConvertToBase(125, lngMaxID)
        UsedIDList.Add(strMaxID)
        If strMaxID.Length > PMaxIDLength Then
            PMaxIDLength = strMaxID.Length
            My.Settings.MaxIDLength = PMaxIDLength
            My.Settings.Save()
        End If

        Return strMaxID
    End Function

    Public Sub Save(FileName_IDList As String) Implements IListOfIDs.Save
        If Not Directory.Exists(Path.GetDirectoryName(FileName_IDList)) Then
            Dim unused = Directory.CreateDirectory(Path.GetDirectoryName(FileName_IDList))
        End If
        Dim TestFileStream As Stream = File.Create(FileName_IDList)
        Dim serializer As New BinaryFormatter
        serializer.Serialize(TestFileStream, Me)
        TestFileStream.Close()
        pFileName = FileName_IDList
    End Sub

    Public Sub Save() Implements IListOfIDs.Save
        If pFileName.Length > 0 Then
            Dim TestFileStream As Stream = File.Create(pFileName)
            Dim serializer As New BinaryFormatter
            serializer.Serialize(TestFileStream, Me)
            TestFileStream.Close()
        Else
            Dim unused = MsgBox("Can't save. IDList FileName not set yet")
        End If
    End Sub

    Public Function ConvertToBase(nbase As Integer, ByVal num As BigInteger, Optional intMinDigits As Integer = 2) As String Implements IListOfIDs.ConvertToBase
        Dim chars As String
        Dim r As BigInteger
        Dim newNumber As String
        Dim maxBase As Integer
        Dim i As Integer

        'chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõöøùúûüýþÿŒœŠšŸŽžƒ"
        chars = "0123456789aAáÁàÀâÂäÄãÃåÅæÆbBcCçÇdDðÐeEéÉèÈêÊëËfFƒgGhHIIíÍìÌîÎïÏjJkKlLmMnNñÑoOóÓòÒôÔöÖõÕøØœŒpPqQrRsSšŠßtTþÞuUúÚùÙûÛüÜvVwWxXyYýÝÿŸzZžŽ"
        maxBase = Len(chars)

        ' check if we can convert to this base
        If nbase > maxBase Then
            ConvertToBase = ""
        Else

            ' in r we have the offset of the char that was converted to the new base
            newNumber = ""
            While num >= nbase
                r = num Mod nbase
                newNumber = Mid(chars, r + 1, 1) & newNumber
                num /= nbase
            End While

            newNumber = Mid(chars, num + 1, 1) & newNumber

            For i = 1 To Len(newNumber) Mod intMinDigits
                newNumber = 0 & newNumber
            Next i

            ConvertToBase = newNumber
        End If
    End Function

    Public Function ConvertToDecimal(nbase As Integer, ByVal strBase As String) As BigInteger Implements IListOfIDs.ConvertToDecimal
        Dim chars As String
        Dim i As Integer
        Dim intLoc As Integer
        Dim lngTmp As BigInteger

        'chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõöøùúûüýþÿŒœŠšŸŽžƒ"
        chars = "0123456789aAáÁàÀâÂäÄãÃåÅæÆbBcCçÇdDðÐeEéÉèÈêÊëËfFƒgGhHIIíÍìÌîÎïÏjJkKlLmMnNñÑoOóÓòÒôÔöÖõÕøØœŒpPqQrRsSšŠßtTþÞuUúÚùÙûÛüÜvVwWxXyYýÝÿŸzZžŽ"
        lngTmp = 0

        For i = 1 To Len(strBase)
            lngTmp *= nbase
            intLoc = InStr(chars, Mid(strBase, i, 1))
            lngTmp += intLoc - 1
        Next i

        ConvertToDecimal = lngTmp
    End Function

    Private Function CustomFieldID_GetValue(objItem As Object, ByVal UserDefinedFieldName As String) As String
        Dim OlMail As [MailItem]
        Dim OlTask As [TaskItem]
        Dim OlAppt As [AppointmentItem]
        Dim objProperty As [UserProperty]


        If objItem Is Nothing Then
            Return ""
        ElseIf TypeOf objItem Is [MailItem] Then
            OlMail = objItem
            objProperty = OlMail.UserProperties.Find(UserDefinedFieldName)

        ElseIf TypeOf objItem Is [TaskItem] Then
            OlTask = objItem
            objProperty = OlTask.UserProperties.Find(UserDefinedFieldName)
        ElseIf TypeOf objItem Is [AppointmentItem] Then
            OlAppt = objItem
            objProperty = OlAppt.UserProperties.Find(UserDefinedFieldName)
        Else
            objProperty = Nothing
            Dim unused = MsgBox("Unsupported object type")
        End If

        Return If(objProperty Is Nothing,
            "",
            If(IsArray(objProperty.Value), FlattenArry(objProperty.Value), DirectCast(objProperty.Value, String)))

        OlMail = Nothing
        OlTask = Nothing
        OlAppt = Nothing
        objProperty = Nothing

    End Function

    Public Function FlattenArry(varBranch() As Object) As String Implements IListOfIDs.FlattenArry
        Dim i As Integer
        Dim strTemp As String

        strTemp = ""

        For i = 0 To UBound(varBranch)
            strTemp = If(IsArray(varBranch(i)), strTemp & ", " & FlattenArry(varBranch(i)), DirectCast(strTemp & ", " & varBranch(i), String))
        Next i
        If strTemp.Length <> 0 Then strTemp = Right(strTemp, Len(strTemp) - 2)
        FlattenArry = strTemp
    End Function
End Class

File: ToDoModel/Dimension Libraries/ID/ListOfIDs.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Numerics;
using System.Runtime.Serialization.Formatters.Binary;
using System.Windows.Forms;
using Microsoft.Office.Interop.Outlook;
using Outlook = Microsoft.Office.Interop.Outlook;


using UtilitiesVB;

namespace ToDoModel
{

    [Serializable()]
    public class ListOfIDs : IListOfIDs
    {

        private List<string> _usedIDList;
        private long _maxIDLength;
        private string _filepath = "";

        public ListOfIDs(List<string> listUsedID)
        {
            UsedIDList = listUsedID;
        }

        public ListOfIDs(string FilePath, Outlook.Application OlApp)
        {
            LoadFromFile(FilePath: FilePath, OlApp: OlApp);
        }

        public ListOfIDs()
        {
            _usedIDList = new List<string>();
        }

        public static ListOfIDs LoadFromFile(string FilePath, Outlook.Application OlApp)
        {
            var tmpIDList = new ListOfIDs();

            if (File.Exists(FilePath))
            {
                var deserializer = new BinaryFormatter();
                try
                {
                    using (Stream TestFileStream = File.OpenRead(FilePath))
                    {
                        tmpIDList = (ListOfIDs)deserializer.Deserialize(TestFileStream);
                    }
                }

                catch (UnauthorizedAccessException ex)
                {
                    tmpIDList = ProcessFileError(OlApp, "Unexpected File Access Error. Recreate the list?");
                }

                catch (IOException ex)
                {
                    tmpIDList = ProcessFileError(OlApp, "Unexpected IO Error. Is IDList File Corrupt?");
                }

                catch (InvalidCastException ex)
                {
                    tmpIDList = ProcessFileError(OlApp, "File exists but cannot cast to ListOfIDs. Recreate the list?");
                }
            }

            else
            {
                tmpIDList = ProcessFileError(OlApp, "File " + FilePath + " does not exist. Recreate the List?");
            }

            tmpIDList.Filepath = FilePath;
            return tmpIDList;
        }

        private static ListOfIDs ProcessFileError(Outlook.Application OlApp, string msg)
        {
            var tmpIDList = new ListOfIDs();
            var result = MessageBox.Show(msg, "Error",MessageBoxButtons.YesNo);
            if (result == DialogResult.Yes)
            {
                tmpIDList.RefreshIDList(OlApp);
            }
            else
            {
                MessageBox.Show("Returning an empty list of ToDoIDs");
            }
            return tmpIDList;
        }

        public void RefreshIDList(Outlook.Application Application)
        {
            var unused = new object();
            var _dataModel = new TreeOfToDoItems();
            List<object> _toDoList;
            UsedIDList = new List<string>();

            _toDoList = _dataModel.GetToDoList(TreeOfToDoItems.LoadOptions.vbLoadAll, Application);

            foreach (object _objItem in _toDoList)
            {
                string strID = CustomFieldID_GetValue(_objItem, "ToDoID");
                if (UsedIDList.Contains(strID) == false & strID.Length != 0)
                {
                    UsedIDList.Add(strID);
                    if (strID.Length > _maxIDLength)
                        _maxIDLength = strID.Length;
                }
            }
        }

        /// <summary>
    /// Function Invokes the DataModel_ToDoTree.ReNumberIDs() method at the root level which 
    /// recursively calls DataModel_ToDoTree.ReNumberChildrenIDs() and then invokes the
    /// ListOfIDs.Save() Method
    /// </summary>
    /// <param name="OlApp">Pointer to Outlook Application</param>
        public void CompressToDoIDs(Outlook.Application OlApp)
        {
            var _dataModel = new TreeOfToDoItems();
            _dataModel.LoadTree(TreeOfToDoItems.LoadOptions.vbLoadAll, OlApp);
            _dataModel.ReNumberIDs(this);
        }

        public long MaxIDLength
        {
            get
            {
                if (_maxIDLength == 0L)
                {
                    long maxLen = 0L;
                    foreach (string strID in UsedIDList)
                    {
                        if (strID.Length > maxLen)
                        {
                            maxLen = strID.Length;
                        }
                    }
                    _maxIDLength = maxLen;
                }
                return _maxIDLength;

            }
        }

        public List<string> UsedIDList
        {
            get
            {
                return _usedIDList;
            }
            set
            {
                _usedIDList = value;
            }
        }

        public string Filepath
        {
            get
            {
                return _filepath;
            }
            set
            {
                _filepath = value;
            }
        }

        public string GetNextAvailableToDoID(string strSeed)
        {
            bool blContinue = true;
            var lngMaxID = ConvertToDecimal(125, strSeed);
            string strMaxID = "";

            while (blContinue)
            {
                lngMaxID += 1;
                strMaxID = ConvertToBase(125, lngMaxID);
                if (UsedIDList.Contains(strMaxID) == false)
                {
                    blContinue = false;
                }
            }
            UsedIDList.Add(strMaxID);
            if (strMaxID.Length > _maxIDLength)
            {
                _maxIDLength = strMaxID.Length;
                Properties.Settings.Default.MaxIDLength = (int)_maxIDLength;
                Properties.Settings.Default.Save();
            }
            return strMaxID;
        }

        public string GetMaxToDoID()
        {
            string strMaxID = UsedIDList.Max();
            var lngMaxID = ConvertToDecimal(125, strMaxID);
            lngMaxID += 1;
            strMaxID = ConvertToBase(125, lngMaxID);
            UsedIDList.Add(strMaxID);
            if (strMaxID.Length > _maxIDLength)
            {
                _maxIDLength = strMaxID.Length;
                Properties.Settings.Default.MaxIDLength = (int)_maxIDLength;
                Properties.Settings.Default.Save();
            }

            return strMaxID;
        }

        public void Save(string Filepath)
        {
            if (!Directory.Exists(Path.GetDirectoryName(Filepath)))
            {
                Directory.CreateDirectory(Path.GetDirectoryName(Filepath));
            }

            var serializer = new BinaryFormatter();
            using (Stream TestFileStream = File.Create(Filepath))
            {
                serializer.Serialize(TestFileStream, this);
            }

            this.Filepath = Filepath;
        }

        public void Save()
        {
            if (Filepath.Length > 0)
            {
                var serializer = new BinaryFormatter();
                using (Stream TestFileStream = File.Create(Filepath))
                {
                    serializer.Serialize(TestFileStream, this);
                }
            }
            else
            {
                MessageBox.Show("Can't save. IDList FileName not set yet");
            }
        }

        public string ConvertToBase(int nbase, BigInteger num, int intMinDigits = 2)
        {
            string ConvertToBaseRet = default;
            string chars;
            BigInteger r;
            string newNumber;
            int maxBase;
            int i;

            // chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõöøùúûüýþÿŒœŠšŸŽžƒ"
            chars = "0123456789aAáÁàÀâÂäÄãÃåÅæÆbBcCçÇdDðÐeEéÉèÈêÊëËfFƒgGhHIIíÍìÌîÎïÏjJkKlLmMnNñÑoOóÓòÒôÔöÖõÕøØœŒpPqQrRsSšŠßtTþÞuUúÚùÙûÛüÜvVwWxXyYýÝÿŸzZžŽ";
            maxBase = (chars.Length);

            // check if we can convert to this base
            if (nbase > maxBase)
            {
                ConvertToBaseRet = "";
            }
            else
            {

                // in r we have the offset of the char that was converted to the new base
                newNumber = "";
                while (num >= nbase)
                {
                    r = num % nbase;
                    newNumber = chars.Substring((int)(r + 1), 1) + newNumber;
                    num /= nbase;
                }

                newNumber = chars.Substring((int)(num + 1), 1) + newNumber;

                var loopTo = (newNumber.Length) % intMinDigits;
                for (i = 1; i <= loopTo; i++)
                    newNumber = 0 + newNumber;

                ConvertToBaseRet = newNumber;
            }

            return ConvertToBaseRet;
        }

        public BigInteger ConvertToDecimal(int nbase, string strBase)
        {
            BigInteger ConvertToDecimalRet = default;
            string chars;
            int i;
            int intLoc;
            BigInteger lngTmp;

            // chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõöøùúûüýþÿŒœŠšŸŽžƒ"
            chars = "0123456789aAáÁàÀâÂäÄãÃåÅæÆbBcCçÇdDðÐeEéÉèÈêÊëËfFƒgGhHIIíÍìÌîÎïÏjJkKlLmMnNñÑoOóÓòÒôÔöÖõÕøØœŒpPqQrRsSšŠßtTþÞuUúÚùÙûÛüÜvVwWxXyYýÝÿŸzZžŽ";
            lngTmp = 0;

            var loopTo = (strBase.Length -1);
            for (i = 1; i <= loopTo; i++)
            {
                lngTmp *= nbase;
                intLoc = chars.IndexOf(strBase.Substring(i, 1));
                lngTmp += intLoc - 1;
            }

            ConvertToDecimalRet = lngTmp;
            return ConvertToDecimalRet;
        }

        private string CustomFieldID_GetValue(object objItem, string UserDefinedFieldName)
        {
            MailItem OlMail;
            TaskItem OlTask;
            AppointmentItem OlAppt;
            UserProperty objProperty;


            if (objItem is null)
            {
                return "";
            }
            else if (objItem is MailItem)
            {
                OlMail = (MailItem)objItem;
                objProperty = OlMail.UserProperties.Find(UserDefinedFieldName);
            }

            else if (objItem is TaskItem)
            {
                OlTask = (TaskItem)objItem;
                objProperty = OlTask.UserProperties.Find(UserDefinedFieldName);
            }
            else if (objItem is AppointmentItem)
            {
                OlAppt = (AppointmentItem)objItem;
                objProperty = OlAppt.UserProperties.Find(UserDefinedFieldName);
            }
            else
            {
                objProperty = null;
                MessageBox.Show("Unsupported object type");
            }

            return objProperty is null ? "" : objProperty is Array ? FlattenArry((object[])objProperty.Value) : (string)objProperty.Value;

            OlMail = null;
            OlTask = null;
            OlAppt = null;
            objProperty = null;

        }

        public string FlattenArry(object[] varBranch)
        {
            string FlattenArryRet = default;
            int i;
            string strTemp;

            strTemp = "";

            var loopTo = (varBranch.Length - 1);
            for (i = 0; i <= loopTo; i++)
                strTemp = varBranch[i] is Array ? strTemp + ", " + FlattenArry((object[])varBranch[i]) : strTemp + ", " + varBranch[i];
            if (strTemp.Length != 0)
                strTemp = strTemp.Substring(2);
            FlattenArryRet = strTemp;
            return FlattenArryRet;
        }
    }
}